src/math/lp/nla_core.cpp

```diff
@@ -1809,18 +1838,230 @@ bool core::improve_bounds() {
+            is_strict = !bound_value.y.is_zero();
+            if (k.is_pos())
+                add_lower_bound_monic(non_fixed, bound_value.x / k, is_strict, lambda);
+            else
+                add_upper_bound_monic(non_fixed, bound_value.x / k, is_strict, lambda);
+        }
+        
+        if (upper_bound_is_available(monic_var)) {
+            bound_value = lra.column_upper_bound(monic_var);
+            is_strict = !bound_value.y.is_zero();
+            auto lambda = [vars, monic_var, non_fixed, &lps]() {
+                u_dependency* dep = lps.get_column_upper_bound_witness(monic_var);
+                for (auto v : vars) {
+                    if (v != non_fixed) {
+                        dep = lps.join_deps(dep, lps.get_bound_constraint_witnesses_for_column(v));
+                    }
+                }
+                return dep;
+            };
+            if (k.is_neg())
+                add_lower_bound_monic(non_fixed, bound_value.x / k, is_strict, lambda);
+            else
+                add_upper_bound_monic(non_fixed, bound_value.x / k, is_strict, lambda);
+        }
+    }
+
+    void core::propagate_monic_with_all_fixed(lpvar monic_var, const svector<lpvar>& vars, const rational& k) {
+        auto* lps = &lra;
+        auto lambda = [vars, lps]() { return lps->get_bound_constraint_witnesses_for_columns(vars); };
+        add_lower_bound_monic(monic_var, k, false, lambda);
+        add_upper_bound_monic(monic_var, k, false, lambda);
+    }
+
+    void core::add_bounds_for_zero_var(lpvar monic_var, lpvar zero_var) {
+        auto* lps = &lra;
+        auto lambda = [zero_var, lps]() {
+            return lps->get_bound_constraint_witnesses_for_column(zero_var);
+        };
+        TRACE("add_bound", lra.print_column_info(zero_var, tout) << std::endl;);
+        add_lower_bound_monic(monic_var, lp::mpq(0), false, lambda);
+        add_upper_bound_monic(monic_var, lp::mpq(0), false, lambda);
+    }
+    
+    void core::propagate_monic_non_fixed_with_lemma(lpvar monic_var, const svector<lpvar>& vars, lpvar non_fixed, const rational& k) {
+        lp::impq bound_value;
+        new_lemma lemma(*this, "propagate monic with non fixed");
+        // using += to not assert thath the inequality does not hold
+        lemma += ineq(term(rational(1), monic_var, -k, non_fixed), llc::EQ, 0);
+        lp::explanation exp;
+        for (auto v : m_emons[monic_var].vars()) {
+            if (v == non_fixed) continue;
+            u_dependency* dep = lra.get_column_lower_bound_witness(v);
+            for (auto ci : lra.flatten(dep)) {
+                exp.push_back(ci);
+            }
+            dep = lra.get_column_upper_bound_witness(v);
+            for (auto ci : lra.flatten(dep)) {
+                exp.push_back(ci);
+            }
+        }
+        lemma &= exp;
+    }
+
+    void core::calculate_implied_bounds_for_monic(lp::lpvar monic_var) {
+        m_propagated.reserve(monic_var + 1, false);
```

Comment:
I have added a flag to the monic class itself.
replace this code by:
auto& monic = m_emons[monic_var]

monic.is_propagated() instead of m_propagated[monic_var]

m_emons.set_propagated(monic); insetad of set_bitvector_trail.